packages/@aws-cdk/aws-codepipeline-actions/lib/s3/deploy-action.ts

```diff
@@ -62,7 +79,205 @@ export class S3DeployAction extends Action {
         BucketName: this.props.bucket.bucketName,
         Extract: this.props.extract === false ? 'false' : 'true',
         ObjectKey: this.props.objectKey,
+        CannedACL: this.props.cannedAcl,
+        CacheControl: this.props.cacheControl && this.props.cacheControl.toString(),
       },
     };
   }
 }
+
+/**
+ * Helper class generating RFC 2616 compliant HTTP Cache-Control request or response header values
+ *
+ * @see https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9 RFC 2616 14.9 Cache-Control
+ * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control MDN: Cache-Control
+ */
+export class CacheControlDirective {
```

Comment:
I don't think this calls for such an API, seems to me that a simple "options" structure should be sufficient:

```ts
export interface CacheControlOptions {
  readonly maxAge?: Duration;
  readonly noCache?: boolean;
  readonly noStore?: boolean;
  readonly noTransform?: boolean;
}
```

Reply:
There are two issues I could see with using an interface:
* `max-stale` has an optional duration:
https://github.com/aws/aws-cdk/blob/d4d81dfaf594b636801cb629027b99395ed518c8/packages/%40aws-cdk/aws-codepipeline-actions/lib/s3/deploy-action.ts#L161
Correct me if I'm wrong, but this would require two properties to avoid an union type? Something like `maxStale?: boolean`, and `maxStaleDuration?: Duration`, which would require a XOR check. Possibly a `maxStale?: {duration:? Duration}` object, but it's not particularly elegant either
* The user can't control the order in which the directives are serialized, but I haven't found anything in the RFC that indicates that would be a problem. This could be an issue with specific HTTP clients (browser) implementations?

If neither of these are problematic, I'll remove the builder pattern

Reply:
Looks good to me. Order doesn’t seem to be an issue, and maxStaleDuration makes sense (by the way, maxStale:true and maxStaleDuration:x is fine, the only illegal combination is maxStaleDurarion set with maxStale:false.

Reply:
Got it, I'll try to do it this week 👍

Reply:
Done